### ISSUE TITLE
`IDE0004` false positive when `object` is cast to `dynamic`

### ISSUE THREAD

**Issue by RenderMichael** [dotnet/roslyn#71695]
Some cases when `IDE0004` is triggered when it shouldn't (without it would be a compilation error).

**Version Used**: 

main branch on Sharplab

**Steps to Reproduce**:

``` csharp
T? M<T>(object? o)
{
    return (dynamic?)o; // fine
}

class C
{
    public T? M<T>(object? o)
    {
        return (dynamic?)o; // diagnostic
        
        T1? V<T1>(object? o1)
        {
            return (dynamic?)o1; // fine
        }
    }

    public T M2<T>(object o)
    {
        return (dynamic)o; // diagnostic
    }
    
    public T? M3<T>(object o)
    {
        return (dynamic?)o; // fine when nullability changes
    }
}
```

**Diagnostic Id**:

`"IDE0004"`

**Expected Behavior**:

no diagnostic fired

**Actual Behavior**:

diagnostic fired
